### ISSUE TITLE
[feature]: Filters (Layered Navigation) - Add keyboard focus and the way tab order should work

### ISSUE THREAD

**Issue by dpatil-magento** [magento/pwa-studio#1222]
As part of [#797](https://github.com/magento/pwa-studio/pull/797) - Layered navigation UI feature is added. But the keyboard focus and tab order does not work as expected.

Per meeting with [@soumya-ashok](https://github.com/soumya-ashok), these are the changes we've identified that need to be made:
* Focus on close button first
* “Escape” should always close modal no matter what depth/element is focused, should remember filter state
* Add outline/border to focused elements
* Add better aria labels to actionable for voiceover (get jimothy’s input)
* Kill tooltip or remove a11y arrow key from tooltip ie aria-ignore.
* Using space to select/unselect a color moves focus back to first color. Should remain on selected color/element. Apply treatment to other filter items
* Shift-tab on close should move focus to apply filters/resetfilters/ or last filter category. “Tab/focus trapping”
* On “apply filters” focus should move back to filter button
* On “reset filters” focus should move to first filter category button *not* the close button.

Additional:
* Make width 360px.
* Make expand from left instead of bottom.

Helpful Links:
* <https://webaim.org/techniques/keyboard/>

**Comment by dpatil-magento** [comment]
If there are no tickets for Mini-cart, Main menu sign-in section then those can also be considered as part of this ticket.

**Comment by brendanfalkowski** [comment]
Adding some notes from my research in this area — specific to accessibility:

A. When focus shifts to a modal/panel, it can be useful for the first element receiving focus to be a "close" button. This makes it obvious to the user they can advance through that content or reverse from the trapped area.

B. When trapping focus in a modal/panel make sure the URL bar is still accessible in the tabbing order.

**Comment by sirugh** [comment]
> If there are no tickets for Mini-cart, Main menu sign-in section then those can also be considered as part of this ticket.

I disagree. There are large tickets in progress for refactoring those components and I think we should separate the a11y concerns for each. For now this ticket should apply to the filter modal/view only.

**Comment by sirugh** [comment]
Discussed with team but this will basically be blocked on #1432. It's just not a high priority right now as we have other modals that are closer to being refactored and functional.

**Comment by awilcoxa** [comment]
Marked as P4S4 low priority during grooming, will re-address if this becomes a larger issue with community.

**Comment by dani97** [comment]
@magento I am working on it

**Comment by sirugh** [comment]
This ticket was being worked on but the original contributor had other demands of their time so for now it is moving to the backlog. 

Whomever wishes to pick this up, please reference existing the existing PR, #2787, that may contain relevant context about the desired solution.

**Comment by davemacaulay** [comment]
This is being tracked internally at https://jira.corp.magento.com/browse/PWA-174

**Comment by vasilii-b** [comment]
@magento I am working on it

**Comment by sirugh** [comment]
@vasilii-b thanks for taking a look at this! Please reference the prior history, specifically any discussion from #2787

**Comment by vasilii-b** [comment]
Thanks for the suggestion @sirugh.
I've read the thread and went down to #2843. Do you have any insights about that PR being closed?

I was researching what other library can we use for this scope, so we don't end up in a bunch of code written again and again for each modal on the website, and ended up with https://www.npmjs.com/package/focus-trap-react.

To my understanding, `react-aria` and `focus-trap-react` are very similar.

Before implementing any solution, I'd like to make sure I do it using the right tools.

Should this topic (a general approach on handling focus in modals) be groomed or it was already?

Thank you!

**Comment by sirugh** [comment]
> I've read the thread and went down to #2843. Do you have any insights about that PR being closed?

If you mean the spike, we just closed it because it was a research spike to evaluate feasibility. If you want to use it as a baseline for your work, that would be acceptable.

> To my understanding, react-aria and focus-trap-react are very similar.

Using `react-aria` would be preferable, especially given that it is an Adobe library and Magento is part of Adobe :)

> Should this topic (a general approach on handling focus in modals) be groomed or it was already?

I want to say that we are slowly consolidating around a single use of modal/dialog but this specific issue is targeting the left nav which is implemented a bit differently (I think). If you, while working on this issue, find a solution that would be generic enough to apply to the other dialogs in the app, please open a new feature request with details and I'll bring it up with the internal team. For now though, please just focus on a11y in the left nav.

**Comment by vasilii-b** [comment]
Thanks for the details, @sirugh!
I'll continue with a11y for now, as suggested.
Thank you!

**Comment by schensley** [comment]
@vasilii-b This looks really good, but there is one thing that doesn't behave as requested...

> On “apply filters” focus should move back to filter button 

Instead when the filters are applied (and the drawer closed) the user will find themselves back at the top of the page, having to tab, tab, tab back to the Filters button. (adding a video for reference).

Otherwise things look good.

**Comment by vasilii-b** [comment]
Thanks for noticing that @schensley.
Seems I missed it, my bad.

I'll look into it

**Comment by sirugh** [comment]
@vasilii-b hey are you still working on this? Looks like we're just waiting for that last fix before we move this to review.

**Comment by vasilii-b** [comment]
Hi @sirugh ,
I was working on the fix for the mentioned bugs but had some difficulties with handling them.
Whom I can contact for a hand of help?
Thanks in advance.

**Comment by sirugh** [comment]
Push whatever the latest state is for your changes and let me know and I'll try to look next week. We're out Monday but will be back Tuesday.

**Comment by vasilii-b** [comment]
Thanks, @sirugh!

What I have achieved so far in the scope of bug fixing - set the focus to the "Filters" button after there is applied some filters that are already in the cache. Didn't figure it out how to do it for the case when the category component is re-rendered completely.

Please let me know what I did wrong.

**Comment by sirugh** [comment]
> Instead when the filters are applied (and the drawer closed) the user will find themselves back at the top of the page, having to tab, tab, tab back to the Filters button. (adding a video for reference).


>  Didn't figure it out how to do it for the case when the category component is re-rendered completely.

The app currently re-renders the entire `Category` component when new filters are applied. Once it makes the new query for data, it caches those results, which is why sometimes everything is unmounted and other times it seems to only update the result set.

The auto-focus of the `Filter` button works properly when there is cached data because the `FocusScope` can return focus back to the button. When there is no cached data, everything unmounts so the button "loses" focus.

I see two ways around this:

1) Automatically focus the Filter button _every_ time `Category` data loads (Easy, but probably not preferred.
2) Persist the Filter button. (preferred, but more complex). I have pushed a PR for you [here](https://github.com/vasilii-b/pwa-studio/pull/1). Please let me know your thoughts.

Regardless of how you proceed you will need to also make sure to apply the changes/fixes to the [Search page](https://github.com/magento/pwa-studio/blob/feature/1222-layered-nav-add-keyboard-focus-and-a11y-rugh-edit/packages/venia-ui/lib/components/SearchPage/searchPage.js#L78-L96) as it also uses a filter modal and button.

**Comment by vasilii-b** [comment]
Thanks for the help here @sirugh.
Changes have been added.
